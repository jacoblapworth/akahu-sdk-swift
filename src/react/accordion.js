import XUIAccordion from './components/accordion/XUIAccordion';

export default XUIAccordion;
