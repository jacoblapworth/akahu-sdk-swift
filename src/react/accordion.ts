import XUIAccordion from './components/accordion/XUIAccordion';
import XUIAccordionItem from './components/accordion/XUIAccordionItem';

export { XUIAccordion as default, XUIAccordion, XUIAccordionItem };
