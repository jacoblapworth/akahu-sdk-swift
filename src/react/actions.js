import XUIActions from './components/actions/XUIActions';

export default XUIActions;
