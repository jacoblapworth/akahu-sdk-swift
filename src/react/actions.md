<div class="xui-margin-vertical">
	<a href="../section-components-controls-button.html#components-controls-button-12" isDocLink>Actions in the XUI Documentation</a>
</div>

Actions are used to wrap a selection of buttons (often two) and provide formatting. They can be embedded in a XUIModal, in a XUIPageHeader, in the footer of a XUIPanel, and other places. These are more generic than Banner Actions or Toast Actions.

#### Default action

A default action aligns a primary and secondary button to the right, with the primary as the rightmost.

```jsx harmony
import XUIActions from '@xero/xui/react/actions';
import XUIButton from '@xero/xui/react/button';

<XUIActions
  primaryAction={<XUIButton variant="main">Continue</XUIButton>}
  secondaryAction={<XUIButton>Cancel</XUIButton>}
/>;
```

#### Linear action

The linear variant aligns the primary button to the right and the secondary button to the left.

```jsx harmony
import XUIActions from '@xero/xui/react/actions';
import XUIButton from '@xero/xui/react/button';

<XUIActions
  isLinear
  primaryAction={<XUIButton variant="main">Next</XUIButton>}
  secondaryAction={<XUIButton>Previous</XUIButton>}
/>;
```
