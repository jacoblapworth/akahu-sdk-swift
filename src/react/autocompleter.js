import Autocompleter from './components/autocompleter/Autocompleter';
import AutocompleterInput from './components/autocompleter/AutocompleterInput';
import SecondarySearch from './components/autocompleter/SecondarySearch';
import EmptyState from './components/autocompleter/EmptyState';
import { boldMatch, decorateSubStr } from './components/autocompleter/helpers/highlighting';

export {
	Autocompleter as default,
	AutocompleterInput,
	EmptyState,
	SecondarySearch,
	boldMatch,
	decorateSubStr
};
