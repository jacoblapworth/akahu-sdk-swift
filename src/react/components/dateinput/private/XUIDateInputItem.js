import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import dateIcon from '@xero/xui-icon/icons/date-end';
import dateStartIcon from '@xero/xui-icon/icons/date-start';
import { parseDate, DateFormat } from '@xero/blind-date';
import { v4 as uuidv4 } from 'uuid';
import cn from 'classnames';
import { ns } from '../../helpers/xuiClassNamespace';

import {
  XUIDropdownFooter,
  XUIDropdownPanel,
  XUIDropdownToggled,
  XUINestedDropdown,
} from '../../../dropdown';
import XUIDatePicker from '../../../datepicker';
import XUIPicklist, { XUIPickitem } from '../../../picklist';
import XUITextInput, { XUITextInputSideElement } from '../../../textinput';
import { baseSizeClasses } from '../../textinput/private/constants';
import XUIIcon from '../../icon/XUIIcon';
import formatSelectedDateToString from './helpers/formatSelectedDateToString';
import { eventKeyValues } from '../../helpers/reactKeyHandler';
import { isDayOutsideRange } from '../../datepicker/helpers/utils';

/**
 * Keyboard bindings to ignore. Space doesn't select in an autocompleter; left and
 * right arrow keys should move cursor in the input
 *
 * @private
 * @type {Array}
 */
const ignoreKeyboardEvents = [32, 37, 39];
class XUIDateInputItem extends Component {
  state = {
    selectedSuggestedDate: null,
    activePanel: null,
    inputValue: null,
    isDateInvalid: null,
    pickitemInitialHighlight: true,
  };

  ddtRef = createRef(null);

  datepickerRef = createRef(null);

  suggestedDatePanelRef = createRef(null);

  inputRef = createRef(null);

  triggerRef = createRef(null);

  /** Generate id's for panels */
  customDatesId = uuidv4();

  suggestedDatesId = uuidv4();

  suggestedDatesFooterId = uuidv4();

  componentDidMount() {
    this.setState({
      activePanel: this.props.suggestedDates ? this.suggestedDatesId : this.customDatesId,
    });

    this.addListeners();
  }

  componentWillUnmount() {
    this.removeListeners();
  }

  /**
   * Add the global event listeners.
   */
  addListeners = () => {
    if (window) {
      window.addEventListener('keyup', this._keyUpHandler);
    }
  };

  /**
   * Remove any global event listeners associated with a given date input.
   */
  removeListeners = () => {
    if (window) {
      window.removeEventListener('keyup', this._keyUpHandler);
    }
  };

  /**
   * @param {Object} event - A keyUp event which is set to be listened to by componentDidMount.
   * For DatePicker to close when the user presses the esc key, we need to attach a key
   * listener to a parent (`window` by default) to catch this key press
   */
  _keyUpHandler = event => {
    if (event.key === eventKeyValues.escape) {
      this.closeDropdown();
      this.inputRef?.current.focus();
    }

    if (event.key === eventKeyValues.tab && event.target === this.inputRef?.current) {
      this.handleInitialFocus();
    }
  };

  // Can't use this this because when you first interact with DI you open DP and focus DI,
  // but if you press down arrow, you open DP but focus DP instead of DI.
  // afterDdtOpens = focusDatePicker => {
  //   this.inputRef?.current.focus();
  // };

  closeDropdown = () => {
    this.ddtRef?.current.closeDropdown();
    this.setState({ pickitemInitialHighlight: true });
  };

  onInputChange = event => {
    this.props.onInputChange
      ? this.props.onInputChange(event)
      : (event.persist(), this.setState({ inputValue: event.target.value }));
  };

  focusDatePicker = () => {
    if (this.state.activePanel === this.customDatesId) {
      this.datepickerRef.current.focus();
    }
  };

  handleInitialFocus = () => {
    if (!this.props.isDisabled) {
      this.setState({ pickitemInitialHighlight: false });
      this.ddtRef.current.openDropdown(() => {
        // This has to be in a timeout until Dropdown is extended to open a panel without focusing it.
        // DropdownPanel will always focus the opened panel so this has to happen after.
        // This doesn't solve everything because input focus will flick if input is focused again.
        // onOpenAnimationEnd can't be used here because down arrow interaction requires different behaviour,
        // after opening DD panel, DatePicker has to be focused instead of DateInput like in this case.
        setTimeout(() => {
          this.inputRef?.current.focus();
          this.inputRef?.current.select();
        }, 100);
      });
    }
  };

  // When focused with a down arrow, it should open DP but not focus DI back.
  // When pressing 'esc' you need to focus the input but don't open DP.
  // onInputFocus = event => {
  //   event.target.select();
  // };

  onIconFocus = () => {
    if (!this.props.isDisabled) {
      this.inputRef?.current.focus();
      this.ddtRef.current.openDropdown();
    }
  };

  onInputKeyDown = event => {
    if (event.key === eventKeyValues.enter || event.key === eventKeyValues.tab) {
      this.ddtRef.current.closeDropdown();
      this.setDate();
    }

    if (event.key === eventKeyValues.down) {
      this.ddtRef.current.openDropdown(() => {
        // Focus first item only after pressing down arrow (not when dropdown is visible during initial input focus).
        if (this.props.suggestedDates) {
          this.ddtRef.current.dropdown.highlightFirstItem();
        }

        // This is in a timeout due to DropdownPanel focusing the panel after it's opened.
        setTimeout(() => {
          this.datepickerRef?.current?.focus();

          // focus suggested panel for single input
          if (this.props.suggestedDates) {
            this.suggestedDatePanelRef?.current?.focus();
          }
        }, 100);
      });
    }
  };

  setDate = () => {
    if (!this.state.inputValue) return;

    const { minDate, maxDate } = this.props;
    const base = new Date();
    const now = new Date();
    const dateFormat = DateFormat.DMY;
    const parsedDate = parseDate({ now, base, dateFormat }, this.state.inputValue);

    if (parsedDate && !isDayOutsideRange(parsedDate, minDate, maxDate)) {
      this.onSelectDate(parsedDate, true);
    } else {
      this.setState({
        isDateInvalid: true,
      });
    }
  };

  onSelectDate = (date, isFocusEvent) => {
    this.setState({
      inputValue: null,
      isDateInvalid: false,
      selectedSuggestedDate: 'custom',
    });

    if (!this.props.preventFocusOnSelect) {
      this.inputRef?.current.focus();
    }

    this.props.onSelectDate?.(date);

    if (!isFocusEvent) {
      this.props.closeOnSelect && this.closeDropdown();
    }
  };

  showDatepickerPanel = () => {
    this.setState({
      activePanel: this.customDatesId,
    });
  };

  onDropdownClose = () => {
    this.state.inputValue && this.setDate();
    this.setState({
      activePanel: this.props.suggestedDates ? this.suggestedDatesId : this.customDatesId,
    });
  };

  selectSuggestedDate = selectedSuggestedDateId => {
    const { suggestedDates } = this.props;
    const suggestedDate = suggestedDates.find(cd => cd.id === selectedSuggestedDateId);

    this.setState({
      selectedSuggestedDate: suggestedDate.id,
    });

    this.onSelectDate(suggestedDate.getDate());
  };

  render() {
    const {
      closeOnSelect,
      suggestedDates,
      displayedMonth,
      exposeInputRef,
      hintMessage,
      inputClassName,
      inputFieldClassName,
      inputLabel,
      isDisabled,
      isInvalid,
      locale,
      maxDate,
      minDate,
      onInputChange, // Destructured so as not to spread.
      onSelectDate, // Destructured so as not to spread.
      preventFocusOnSelect,
      selectDateIcon,
      selectDateLabel,
      selectedDate,
      size,
      triggerClassName,
      validationMessage,
      qaHook,
      ...spreadProps
    } = this.props;
    const { activePanel, selectedSuggestedDate, inputValue } = this.state;
    const isDateInvalid = isInvalid || this.state.isDateInvalid;

    /** Trigger where users focus to select date or type in a date shortcut */
    const trigger = (
      <div
        className={cn(
          `${ns}-dateinputitem`,
          triggerClassName,
          isDisabled && `${ns}-dateinputitem-is-disabled`,
        )}
        ref={this.triggerRef}
      >
        <XUITextInput
          autoComplete="off"
          containerClassName={`${ns}-dateinputitem--input`}
          fieldClassName={inputFieldClassName}
          hintMessage={hintMessage}
          inputClassName={inputClassName}
          inputRef={el => {
            if (exposeInputRef) {
              exposeInputRef(el);
            }

            this.inputRef.current = el;
          }}
          isDisabled={isDisabled}
          isInvalid={isDateInvalid}
          label={inputLabel}
          leftElement={
            <XUITextInputSideElement onClick={this.onIconFocus}>
              <XUIIcon color="black" icon={dateIcon} isBoxed />
            </XUITextInputSideElement>
          }
          onBlur={this.setDate}
          onChange={this.onInputChange}
          onClick={this.handleInitialFocus}
          onKeyDown={this.onInputKeyDown}
          qaHook={qaHook && `${qaHook}-dateinputitem--input`}
          size={size}
          validationMessage={validationMessage}
          value={formatSelectedDateToString(selectedDate, inputValue)}
          {...spreadProps}
        />
      </div>
    );

    /** Dropdown footer: */
    const dropdownFooter = (
      <XUIDropdownFooter
        className={`${ns}-dateinputitem-suggesteddates--footer`}
        pickItems={
          <XUIPickitem
            className={`${ns}-dateinputitem-suggesteddates--footer--item`}
            id={this.suggestedDatesFooterId}
            leftElement={<XUIIcon color="black" icon={selectDateIcon || dateStartIcon} isBoxed />}
            onClick={this.showDatepickerPanel}
          >
            {selectDateLabel}
          </XUIPickitem>
        }
      />
    );

    /** The main dropdown */
    const dateInputDropdown = (
      <XUINestedDropdown
        currentPanelId={activePanel}
        ignoreKeyboardEvents={ignoreKeyboardEvents}
        onPanelChange={this.focusDatePicker}
        qaHook={qaHook && `${qaHook}-dateinputitem-dropdown`}
      >
        <XUIDropdownPanel
          allowNarrowViewportAutoHeight
          panelId={this.customDatesId}
          qaHook={qaHook && `${qaHook}-dateinputitem-datepicker`}
        >
          <XUIDatePicker
            displayedMonth={displayedMonth}
            locale={locale}
            maxDate={maxDate}
            minDate={minDate}
            onSelectDate={this.onSelectDate}
            ref={this.datepickerRef}
            selectedDate={selectedDate}
          />
        </XUIDropdownPanel>
        <XUIDropdownPanel
          allowNarrowViewportAutoHeight
          footer={dropdownFooter}
          panelId={this.suggestedDatesId}
          panelRef={el => (this.suggestedDatePanelRef.current = el)}
          qaHook={qaHook && `${qaHook}-dateinputitem-suggesteddates`}
          shouldManageInitialHighlight={this.state.pickitemInitialHighlight}
        >
          <XUIPicklist>
            {suggestedDates?.map(({ id, text }) => (
              <XUIPickitem
                id={id}
                isSelected={selectedSuggestedDate === id}
                key={id}
                onSelect={this.selectSuggestedDate}
                value={id}
              >
                {text}
              </XUIPickitem>
            ))}
          </XUIPicklist>
        </XUIDropdownPanel>
      </XUINestedDropdown>
    );

    return (
      <XUIDropdownToggled
        closeOnSelect={closeOnSelect}
        closeOnTab={false}
        dropdown={dateInputDropdown}
        onClose={this.onDropdownClose}
        ref={this.ddtRef}
        restrictedToViewPort={false}
        trigger={trigger}
        triggerClickAction="none"
      />
    );
  }
}

XUIDateInputItem.propTypes = {
  /** Whether or not the dropdown should automatically be hidden when the user selects something */
  closeOnSelect: PropTypes.bool,

  /**
   * A date which represents the year and month that the calendar will display. Could
   * be any day in the given day and month.
   */
  displayedMonth: PropTypes.instanceOf(Date),

  /** Sets a ref for the input element */
  exposeInputRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),

  /** Hint message to display below input */
  hintMessage: PropTypes.string,

  /** Class names to be added to the input element */
  inputClassName: PropTypes.string,

  /** Class names to be added to the input field wrapper element */
  inputFieldClassName: PropTypes.string,

  /** Input label */
  inputLabel: PropTypes.string,

  /** Whether the input is disabled */
  isDisabled: PropTypes.bool,

  /** Whether the current input value is invalid */
  isInvalid: PropTypes.bool,

  /** The locale of the calendar. Defaults to En */
  locale: PropTypes.string,

  /**
   * If you want to disable every date after a given day, pass in the maximum enabled
   * date here. Can be used with the isDateDisabled function.
   */
  maxDate: PropTypes.instanceOf(Date),

  /**
   * If you want to disable every date before a given day, pass in the minimum enabled
   * date here. Can be used with the isDateDisabled function.
   */
  minDate: PropTypes.instanceOf(Date),

  /** Callback for when the input changes  */
  onInputChange: PropTypes.func,

  /**
   * Callback for when the user selects a date. Will fire even if the date has
   * already been selected.
   */
  onSelectDate: PropTypes.func,

  /** Prevents focusing input after selecting date - allows to focus the second input in date range without an infinite loop. */
  preventFocusOnSelect: PropTypes.bool,

  qaHook: PropTypes.string,

  /**
   * Icon displayed next to calendar selection in suggested dates dropdown.
   */
  selectDateIcon: PropTypes.shape({
    height: PropTypes.number,
    path: PropTypes.string,
    width: PropTypes.number,
  }),

  /** Label for an item opening DatePicker (with suggested date mode) */
  selectDateLabel: PropTypes.string,

  /** Value of the date input. Must be a Date object */
  selectedDate: PropTypes.instanceOf(Date),

  /** Size of the input - Can be `xsmall`, `small` or `medium` */
  size: PropTypes.oneOf(Object.keys(baseSizeClasses)),

  /**
   * An array of objects with suggested dates for users to select quickly.
   * Each object consists of an `id`, display `text` and `getDate` function returning the date to be selected.
   */
  suggestedDates: PropTypes.arrayOf(
    PropTypes.shape({
      getDate: PropTypes.func.isRequired,
      id: PropTypes.string,
      text: PropTypes.string,
    }),
  ),

  /** CSS class(es) to go on the trigger element which contains the input */
  triggerClassName: PropTypes.string,

  /** Message to display below input when invalid date inputted */
  validationMessage: PropTypes.node,
};

XUIDateInputItem.defaultProps = {
  closeOnSelect: true,
  displayedMonth: new Date(),
  locale: 'en',
};

export default XUIDateInputItem;
