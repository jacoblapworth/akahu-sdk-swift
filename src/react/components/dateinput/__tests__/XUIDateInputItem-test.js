import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { screen } from '@testing-library/dom';
import Enzyme, { mount } from 'enzyme';
import { nanoid } from 'nanoid';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import renderer from 'react-test-renderer';
import { axe, toHaveNoViolations } from 'jest-axe';
import XUIDateInputItem from '../private/XUIDateInputItem';

const { renderIntoDocument } = require('react-dom/test-utils');

jest.mock('nanoid');
nanoid.mockImplementation(() => 'testDateinputId');

Enzyme.configure({ adapter: new Adapter() });
expect.extend(toHaveNoViolations);

describe('XUIDateInputItem', () => {
  const selectedDate = new Date(Date.UTC(2020, 11, 15));
  const createComponent = props => (
    <XUIDateInputItem
      inputLabel="Date"
      locale="en-NZ"
      nextButtonAriaLabel="Next month"
      onSelectDate={() => {}}
      prevButtonAriaLabel="Previous month"
      {...props}
      selectedDate={selectedDate}
    />
  );

  it('inserts selected date value', () => {
    const newSelectedDate = new Date(Date.UTC(2020, 11, 18));
    const inputEl = renderer.create(createComponent({ selectedDate: newSelectedDate }));
    expect(inputEl).toMatchSnapshot();
  });

  it('renders with the provided label', () => {
    const inputEl = renderer.create(createComponent({ inputLabel: 'Select date' }));
    expect(inputEl).toMatchSnapshot();
  });

  it('renders with a hint', () => {
    const inputEl = renderer.create(createComponent({ hintMessage: 'hint message' }));
    expect(inputEl).toMatchSnapshot();
  });

  it('renders with validation message', () => {
    const inputEl = renderer.create(
      createComponent({ isInvalid: true, validationMessage: 'validation message' }),
    );
    expect(inputEl).toMatchSnapshot();
  });

  it('renders disabled', () => {
    const inputEl = renderer.create(createComponent({ isDisabled: true }));
    expect(inputEl).toMatchSnapshot();
  });

  it('fires onInputChange callback when text input changed', () => {
    const onInputChange = jest.fn();
    const wrapper = mount(createComponent({ onInputChange }));

    wrapper.find('input').simulate('change', {
      target: {
        value: 'a',
      },
    });

    expect(onInputChange).toHaveBeenCalled();
  });

  it('adds CSS class passed as `triggerClassName`', () => {
    const wrapper = mount(createComponent({ triggerClassName: 'date-input-trigger' }));

    const inputEL = wrapper.find('.date-input-trigger');
    expect(inputEL.length).toEqual(1);
  });

  it('opens DatePicker after focusing the input', () => {
    const wrapper = mount(createComponent({ qaHook: 'date-input-test' }));
    const dropdownOpened = jest.spyOn(wrapper.instance().ddtRef.current, 'openDropdown');

    wrapper.find('input').simulate('click');

    expect(dropdownOpened).toHaveBeenCalled();
  });

  it('opens DatePicker after clicking the icon', () => {
    const wrapper = mount(createComponent({ qaHook: 'date-input-test' }));
    const dropdownOpened = jest.spyOn(wrapper.instance().ddtRef.current, 'openDropdown');

    wrapper.find('.xui-iconwrapper').simulate('click');
    expect(dropdownOpened).toHaveBeenCalled();
  });

  it('should call the passed onFocus handler when focused', () => {
    const dateInputRef = React.createRef();
    render(
      <XUIDateInputItem
        inputLabel="Date"
        locale="en-NZ"
        nextButtonAriaLabel="Next month"
        onSelectDate={() => {}}
        prevButtonAriaLabel="Previous month"
        ref={dateInputRef}
        selectedDate={selectedDate}
      />,
    );

    dateInputRef.current.inputRef.current.focus();
    expect(dateInputRef.current.inputRef.current).toEqual(document.activeElement);
  });

  it('Dropdown opening triggers input focus', () => {
    const wrapper = mount(createComponent({ triggerClassName: 'date-input-trigger' }));
    const dropdownOpened = jest.spyOn(wrapper.instance().ddtRef.current, 'openDropdown');

    wrapper.find('input').simulate('click');

    expect(dropdownOpened).toHaveBeenCalled();
  });

  it('accepts displayedMonth to change DatePicker initial display date', () => {
    const newSelectedDate = new Date(Date.UTC(2020, 1, 15));
    const wrapper = mount(
      createComponent({ selectedDate: newSelectedDate, displayedMonth: newSelectedDate }),
    );

    wrapper.find('input').simulate('click');
    expect(wrapper.find('.xui-datepicker').length).toBe(1);
    const calendarHeadingLabels = wrapper.find('.xui-datepicker--heading-dates label');
    expect(calendarHeadingLabels.length).toBe(2);
    expect(calendarHeadingLabels.first().text()).toEqual('February');
    expect(calendarHeadingLabels.at(1).text()).toEqual('2020');
  });

  it('should pass accessibility testing', async () => {
    const wrapper = mount(createComponent({ inputLabel: 'default label' }));
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });

  test('when a user selects a date then clears the input and presses enter, the selected date should be null', () => {
    // Arrange
    const dateInputRef = React.createRef();
    const onSelectDateCallback = jest.fn();
    render(
      <XUIDateInputItem
        inputLabel="Date"
        locale="en-NZ"
        nextButtonAriaLabel="Next month"
        onSelectDate={onSelectDateCallback}
        prevButtonAriaLabel="Previous month"
        ref={dateInputRef}
        qaHook={'dateinputitem'}
      />,
    );

    const inputNode = screen.getByTestId('dateinputitem-dateinputitem--input--input');

    // Act
    userEvent.type(inputNode, 'Jan 1, 2001{enter}');
    expect(inputNode.value).toBe('Jan 1, 2001');
    userEvent.type(inputNode, '{del}{enter}');

    // Assert
    expect(inputNode.value).toBe('');
  });

  test('when a user selects a date then clears the input and presses enter, the user provided `onSelectDate` prop callback should be fired', () => {
    // Arrange
    const dateInputRef = React.createRef();
    const onSelectDateCallback = jest.fn();
    render(
      <XUIDateInputItem
        inputLabel="Date"
        locale="en-NZ"
        nextButtonAriaLabel="Next month"
        onSelectDate={onSelectDateCallback}
        prevButtonAriaLabel="Previous month"
        ref={dateInputRef}
        qaHook={'dateinputitem'}
      />,
    );

    const inputNode = screen.getByTestId('dateinputitem-dateinputitem--input--input');

    // Act
    userEvent.type(inputNode, 'Jan 1, 2001{enter}');
    expect(inputNode.value).toBe('Jan 1, 2001');
    userEvent.type(inputNode, '{del}{enter}');

    // Assert
    expect(onSelectDateCallback).toBeCalled();
  });
});
