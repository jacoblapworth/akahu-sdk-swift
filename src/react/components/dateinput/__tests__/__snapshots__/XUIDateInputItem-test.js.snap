// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XUIDateInputItem inserts selected date value 1`] = `
<div
  className="xui-dropdown--toggledwrapper"
  data-ref="toggled-wrapper"
  role="presentation"
>
  <div
    aria-activedescendant={null}
    aria-haspopup="listbox"
    className="xui-dateinputitem"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <div
      className="xui-textinputwrapper"
      onKeyDown={[Function]}
      role="presentation"
    >
      <div
        autoComplete="off"
        className="xui-dateinputitem--input xui-textinput xui-textinput-medium"
        onClick={[Function]}
      >
        <div
          className="xui-textinput--sideelement xui-textinput--sideelement-align-top xui-textinput--sideelement-icon"
          onClick={[Function]}
        >
          <div
            className="xui-iconwrapper xui-iconwrapper-medium"
          >
            <svg
              className="xui-icon xui-icon-color-black-faint"
              focusable="false"
              height={13}
              viewBox="0 0 12 13"
              width={12}
            >
              <path
                d="M10 1V0H8v1H4V0H2v1H1C.5 1 0 1.5.01 2L0 12c0 .5.5 1 1 1h10c.5 0 1-.5 1-1V2c0-.5-.5-1-1-1h-1zM1 12V4h10v8H1zm6-4h3v3H7V8z"
                role="presentation"
              />
            </svg>
          </div>
        </div>
        <input
          aria-invalid={null}
          className="xui-textinput--input xui-textinput--input-medium"
          id="xui-testDateinputId-control"
          onBlurCapture={[Function]}
          onChange={[Function]}
          onFocusCapture={[Function]}
          style={
            Object {
              "maxHeight": undefined,
            }
          }
          value="15 Dec 2020"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`XUIDateInputItem renders disabled 1`] = `
<div
  className="xui-dropdown--toggledwrapper"
  data-ref="toggled-wrapper"
  role="presentation"
>
  <div
    aria-activedescendant={null}
    aria-haspopup="listbox"
    className="xui-dateinputitem xui-dateinputitem-is-disabled"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <div
      className="xui-textinputwrapper"
      onKeyDown={[Function]}
      role="presentation"
    >
      <div
        autoComplete="off"
        className="xui-dateinputitem--input xui-textinput xui-textinput-medium xui-textinput-is-disabled"
        onClick={[Function]}
      >
        <div
          className="xui-textinput--sideelement xui-textinput--sideelement-align-top xui-textinput--sideelement-icon"
          onClick={[Function]}
        >
          <div
            className="xui-iconwrapper xui-iconwrapper-medium"
          >
            <svg
              className="xui-icon xui-icon-color-black-faint"
              focusable="false"
              height={13}
              viewBox="0 0 12 13"
              width={12}
            >
              <path
                d="M10 1V0H8v1H4V0H2v1H1C.5 1 0 1.5.01 2L0 12c0 .5.5 1 1 1h10c.5 0 1-.5 1-1V2c0-.5-.5-1-1-1h-1zM1 12V4h10v8H1zm6-4h3v3H7V8z"
                role="presentation"
              />
            </svg>
          </div>
        </div>
        <input
          aria-invalid={null}
          className="xui-textinput--input xui-textinput--input-medium"
          disabled={true}
          id="xui-testDateinputId-control"
          onBlurCapture={[Function]}
          onChange={[Function]}
          onFocusCapture={[Function]}
          style={
            Object {
              "maxHeight": undefined,
            }
          }
          value="15 Dec 2020"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`XUIDateInputItem renders with a hint 1`] = `
<div
  className="xui-dropdown--toggledwrapper"
  data-ref="toggled-wrapper"
  role="presentation"
>
  <div
    aria-activedescendant={null}
    aria-haspopup="listbox"
    className="xui-dateinputitem"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <div
      className="xui-textinputwrapper"
      onKeyDown={[Function]}
      role="presentation"
    >
      <div
        autoComplete="off"
        className="xui-dateinputitem--input xui-textinput xui-textinput-medium"
        onClick={[Function]}
      >
        <div
          className="xui-textinput--sideelement xui-textinput--sideelement-align-top xui-textinput--sideelement-icon"
          onClick={[Function]}
        >
          <div
            className="xui-iconwrapper xui-iconwrapper-medium"
          >
            <svg
              className="xui-icon xui-icon-color-black-faint"
              focusable="false"
              height={13}
              viewBox="0 0 12 13"
              width={12}
            >
              <path
                d="M10 1V0H8v1H4V0H2v1H1C.5 1 0 1.5.01 2L0 12c0 .5.5 1 1 1h10c.5 0 1-.5 1-1V2c0-.5-.5-1-1-1h-1zM1 12V4h10v8H1zm6-4h3v3H7V8z"
                role="presentation"
              />
            </svg>
          </div>
        </div>
        <input
          aria-describedby="xui-testDateinputId-message"
          aria-invalid={null}
          className="xui-textinput--input xui-textinput--input-medium"
          id="xui-testDateinputId-control"
          onBlurCapture={[Function]}
          onChange={[Function]}
          onFocusCapture={[Function]}
          style={
            Object {
              "maxHeight": undefined,
            }
          }
          value="15 Dec 2020"
        />
      </div>
      <div
        className="xui-validation xui-validation-layout"
        id="xui-testDateinputId-message"
        role="status"
      >
        hint message
      </div>
    </div>
  </div>
</div>
`;

exports[`XUIDateInputItem renders with the provided label 1`] = `
<div
  className="xui-dropdown--toggledwrapper"
  data-ref="toggled-wrapper"
  role="presentation"
>
  <div
    aria-activedescendant={null}
    aria-haspopup="listbox"
    className="xui-dateinputitem"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <div
      className="xui-textinputwrapper"
      onKeyDown={[Function]}
      role="presentation"
    >
      <label
        className="xui-text-label xui-fieldlabel-layout"
        htmlFor="xui-testDateinputId-control"
      >
        Select date
      </label>
      <div
        autoComplete="off"
        className="xui-dateinputitem--input xui-textinput xui-textinput-medium"
        onClick={[Function]}
      >
        <div
          className="xui-textinput--sideelement xui-textinput--sideelement-align-top xui-textinput--sideelement-icon"
          onClick={[Function]}
        >
          <div
            className="xui-iconwrapper xui-iconwrapper-medium"
          >
            <svg
              className="xui-icon xui-icon-color-black-faint"
              focusable="false"
              height={13}
              viewBox="0 0 12 13"
              width={12}
            >
              <path
                d="M10 1V0H8v1H4V0H2v1H1C.5 1 0 1.5.01 2L0 12c0 .5.5 1 1 1h10c.5 0 1-.5 1-1V2c0-.5-.5-1-1-1h-1zM1 12V4h10v8H1zm6-4h3v3H7V8z"
                role="presentation"
              />
            </svg>
          </div>
        </div>
        <input
          aria-invalid={null}
          className="xui-textinput--input xui-textinput--input-medium"
          id="xui-testDateinputId-control"
          onBlurCapture={[Function]}
          onChange={[Function]}
          onFocusCapture={[Function]}
          style={
            Object {
              "maxHeight": undefined,
            }
          }
          value="15 Dec 2020"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`XUIDateInputItem renders with validation message 1`] = `
<div
  className="xui-dropdown--toggledwrapper"
  data-ref="toggled-wrapper"
  role="presentation"
>
  <div
    aria-activedescendant={null}
    aria-haspopup="listbox"
    className="xui-dateinputitem"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <div
      className="xui-textinputwrapper"
      onKeyDown={[Function]}
      role="presentation"
    >
      <div
        autoComplete="off"
        className="xui-dateinputitem--input xui-textinput xui-textinput-medium xui-textinput-is-invalid"
        onClick={[Function]}
      >
        <div
          className="xui-textinput--sideelement xui-textinput--sideelement-align-top xui-textinput--sideelement-icon"
          onClick={[Function]}
        >
          <div
            className="xui-iconwrapper xui-iconwrapper-medium"
          >
            <svg
              className="xui-icon xui-icon-color-black-faint"
              focusable="false"
              height={13}
              viewBox="0 0 12 13"
              width={12}
            >
              <path
                d="M10 1V0H8v1H4V0H2v1H1C.5 1 0 1.5.01 2L0 12c0 .5.5 1 1 1h10c.5 0 1-.5 1-1V2c0-.5-.5-1-1-1h-1zM1 12V4h10v8H1zm6-4h3v3H7V8z"
                role="presentation"
              />
            </svg>
          </div>
        </div>
        <input
          aria-describedby="xui-testDateinputId-message"
          aria-invalid={true}
          className="xui-textinput--input xui-textinput--input-medium"
          id="xui-testDateinputId-control"
          onBlurCapture={[Function]}
          onChange={[Function]}
          onFocusCapture={[Function]}
          style={
            Object {
              "maxHeight": undefined,
            }
          }
          value="15 Dec 2020"
        />
      </div>
      <div
        className="xui-validation xui-validation-layout xui-validation-is-invalid"
        id="xui-testDateinputId-message"
        role="status"
      >
        <svg
          className="xui-icon"
          focusable="false"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <path
            d="M8,16 C3.581722,16 0,12.418278 0,8 C0,3.581722 3.581722,0 8,0 C12.418278,0 16,3.581722 16,8 C16,12.418278 12.418278,16 8,16 Z M7,4 L7,9 L9,9 L9,4 L7,4 Z M7,10 L7,12 L9,12 L9,10 L7,10 Z"
            role="presentation"
          />
        </svg>
        validation message
      </div>
    </div>
  </div>
</div>
`;
